src/App.ts

```diff
@@ -0,0 +1,290 @@
+ *                          <- (9) SEND FILE (bin)
+ * (10) RECEIVE FILE (bin) <-
+ */
+export class App
+{
+    private _download_folder = join(process.cwd(), 'downloads');
+    private _transfers_subscription_channel: Function;
+
+    constructor(private _electron: ElectronStarter)
+    {}
+
+    public async listen()
+    {
+        ipcMain.on(start_server_command, this._on_start_server_command.bind(this));
+        ipcMain.on(open_dialog_command, this._on_open_dialog_command.bind(this));
+        ipcMain.on(send_file_command, this._on_send_file_command.bind(this));
+        ipcMain.on(subscribe_to_transfers_command, this._on_subscribe_to_transfers_command.bind(this));
+
+        await this._try_make_download_folder(this._download_folder);
+    }
+
+    private _on_start_server_command(event: IpcMainEvent)
+    {
+        const server = net.createServer(async connectionListener =>
+        {
+            let received_filename = false;
+            let write_stream: WriteStream;
+            let boundary = `|>${nanoid.nanoid()}<|`;
+
+            let filename_buffer = '';
+
+            // (1) Tell the client which is the boundary
+            connectionListener.write(boundary);
+
+            connectionListener.on('close', () =>
+            {
+                received_filename = false;
+                write_stream?.close();
+
+                if (this._transfers_subscription_channel)
+                    this._transfers_subscription_channel(transfer_completed_event);
+            });
+
+            connectionListener.on('data', async data =>
+            {
+                if (!received_filename)
+                {
+                    filename_buffer += data;
+
+                    if (filename_buffer.endsWith(boundary))
+                    {
+                        // (5)-(6) Received the filename + boundary
+                        received_filename = true;
+
+                        const raw_filename = filename_buffer.replace(boundary, '');
+                        const filename = await this._filter_filename(raw_filename);
+
+                        if (this._transfers_subscription_channel)
+                            this._transfers_subscription_channel(transfer_incoming_event, filename);
+
+                        console.log(`Saving file ${filename} into folder ${this._download_folder}`);
+
+                        write_stream = createWriteStream(join(this._download_folder, filename));
+
+                        // (7) Send back the boundary, for the client to know that the server is ready to accept binary data
+                        connectionListener.write(boundary);
+                    }                    
+                }
+                else
+                {
+                    // (10) Receive the binary data
+                    write_stream.write(data);
+                }
+            });
+        });
+
+        server.on('error', err =>
+        {
+            throw err;
```

Comment:
Throwing inside a callback may lead to unexpected behavior 🧑‍🍳

Reply:
Handled exception dispatching visible error to the client side